### PR TITLE
Don't allow uncompressed public keys without prefix 0x04

### PR DESCRIPTION
**Pull request by nlanson**
Was following #520 and through it was a quick fix that I could do:

#### Changes:
- If an uncompressed public key doesn't have prefix 0x04 in `PublicKey::from_slice()`, an error is returned.

<br>

I was wondering if `PublicKey::from_str()` should also enforce the same rules, however I have not incuded this in the PR.

Please let me know if any changes need to be made.

Thanks

### CODE REVIEW

**Comment by Kixunil** [comment]
> I was wondering if `PublicKey::from_str()` should also enforce the same rules

I strongly believe it should.

That being said it seems we have [WET](https://en.wikipedia.org/wiki/Don't_repeat_yourself#WET) code - different deserialization code in `from_slice` and `from_str`. Instead it should be composed by `from_str` decoding the string into slice and then calling `from_slice`. I understand this decoding code is also used in `secp256k1` and I strongly believe the functions for manipulating the encoding should be *some* public API (in `secp256k1` or a separate crate).

This is loosely related to error handling topic where `FromStr::Err` should be an enum that has two variants: invalid encoding and public key error. An [example](https://github.com/LNP-BP/ln-types/blob/master/src/node_id.rs#L311) of how I'd like to do this is in my `ln-types` crate.

Also would strongly prefer to have a special error variant for this case to help people debug this.

**Comment by apoelstra** [comment]
concept ACK, but I agree that we should try to fix the `from_str`/`from_slice` duplication alongside this.

**Comment by nlanson** [comment]
Sounds good, I'm happy to fix the duplication in this PR and will add an error variant for this case.

As for the hex string decoding, there seem to be a few implementations of this across different crates in this org.
- The `secp256k1` crate has  [`from_hex()` function](https://github.com/rust-bitcoin/rust-secp256k1/blob/ef59aea8888d786d12039ffa336f342079a8be45/src/lib.rs#L503) which is used when converting a hex string to a slice in the `PublicKey::from_str()` method.
- The `FromHex` trait in [`bitcoin_hashes` crate](https://github.com/rust-bitcoin/bitcoin_hashes/blob/029d665c657253a33056dbdb1e83bfd9458e6733/src/hex.rs#L60)

which method/trait would be preferred to be used in the `PublicKey::from_str()` method to convert a hex string into a slice in `rust-bitcoin`?

**Comment by apoelstra** [comment]
I'd be okay with either. I think it's fine to use `secp256k1::PublicKey::from_slice` or `secp256k1::PublicKey::from_str` with an explicit preix check.

**Comment by nlanson** [comment]
Just tried adding a new variant to `key::Error`, however came up with a two errors where `From<key::Error`> had been implemented. 

I didn't think it was a good idea to add a new variant to each of the affected error enums that implemented `From<key::Error` so I have got it to convert to a the most related error which in this case was `Secp256k1` variant.

Is there a better way to do this?

**Comment by nlanson** [comment]
I've gone for using the `FromHex` trait to convert hex strings into a slice before passing it onto `PublicKey::from_slice()`.

The conversion uses `Vec::from_hex(s)?` which is not ideal and using `secp256k1::from_hex()` would have better performance since it can be done with an array instead of a vector. However `secp256k1::from_hex()` isn't public so I went with `FromHex` for the time being.

A new variant has also been added to the `key::Error` for handling hex decoding errors.

**Comment by apoelstra** [comment]
I think you could check the string length and use `<[u8; 33]>::from_hex` or `<[u8; 65]>::from_hex` as appropriate.

It's not very clean, but I'm really loath to introduce a new `Vec` allocation, especially when parsing such a tiny object.

**Review by Kixunil — APPROVED** [review]
ACK c8324a688547da08cf3aa4ce1713ece82bc04fc7

File: src/util/base58.rs
Comment (Kixunil) [reply src/util/base58.rs:50]: Great points! The question is do we want to try to do it in this PR or as a part of error handling refactor? I don't see a problem with either approach but if we're going to accept this PR it should probably *not* go to the Taproot release because we'd be removing these variants in the next version.

I believe removing `From<key::Error>` is needed. There's no reason why base58 should have anything to do with key errors.

Also would love to see `base58` as a separate crate with code quality similar to the `base64` crate.

Comment (nlanson) [reply src/util/base58.rs:50]: I had tried to keep the errors in other modules as untouched as possible in this PR but some variants had to be introduced as part of existing implementations. I hadn't actually taken a good look at where the errors could originate from. I will definitely take some time to look at it.

However I think it shouldn't be part of this PR since that changes could be pretty large.

Comment (Kixunil) [reply src/util/base58.rs:50]: Yeah, I'm also a bit more in favor of not in this PR.

Comment (tcharding) [reply src/util/base58.rs:50]: > Also would love to see base58 as a separate crate with code quality similar to the base64 crate.

The base58 crate smash is super easy, I've got it queued up in my tree waiting for the correct time to push.

**Comment by nlanson** [comment]
Oops didn't meant o dismiss your review @Kixunil

**Review by Kixunil — APPROVED** [review]
ACK c0d36efb8b03fbbb4a85331662c06c4640c16292

**Comment by Kixunil** [comment]
@nlanson dismissing happens automatically on push

File: src/util/base58.rs
Comment (Kixunil) [reply src/util/base58.rs]: I ACKed the change but next time I think it'd be better to use GH issues for that to not waste time pushing and reviewing.

Comment (Kixunil) [reply src/util/base58.rs]: > I've got it queued up in my tree

IDK what's your design but I was thinking of also redesigning it slightly to separate encoding and checksum computation so that one can decode the bytes without being forced to compute checksum.

Comment (apoelstra) [reply src/util/base58.rs]: Fine with having a separate `base58` crate but recall we were thinking about having `bitcoin-str` which would combine base58, hex, `rust-bitcoin-bech32` and maybe even `base64-compat`.

Anyway this discussion goes well beyond this PR :)

**Review by apoelstra — APPROVED** [review]
ACK c0d36efb8b03fbbb4a85331662c06c4640c16292

Will wait for a third ack before merging. Agree that the new error variant seems misplaced, but I also don't think there's a good fix without doing a wider error type cleanup.

**Comment by tcharding** [comment]
I'm good with the current changes but purposefully leaving the third ack for someone more experienced in the project to add.

**Comment by apoelstra** [comment]
Let's give it two more days and see if anyone complains.

**Review by sanket1729 — APPROVED** [review]
utACK c0d36efb8b03fbbb4a85331662c06c4640c16292. Not thrilled about the error message expecting len 66, when it can be both 66/130. But can live with it

**Comment by Kixunil** [comment]
Oh, good point about error message! I'm willing to move it to error handling topic but I certainly want it to be addressed.

**Pull request merged**
